indigo/indigo/src/main/scala/indigo/shared/collections/Chain.scala

```diff
@@ -0,0 +1,210 @@
+  given CanEqual[Chain.Wrapped[_], Chain.Empty.type]   = CanEqual.derived
+  given CanEqual[Chain.Wrapped[_], Chain.Singleton[_]] = CanEqual.derived
+  given CanEqual[Chain.Wrapped[_], Chain.Combine[_]]   = CanEqual.derived
+  given CanEqual[Chain.Wrapped[_], Chain.Wrapped[_]]   = CanEqual.derived
+
+  def apply[A](value: A): Chain[A] =
+    Singleton(value)
+
+  def apply[A](values: js.Array[A]): Chain[A] =
+    Wrapped(values)
+
+  def apply[A](values: A*): Chain[A] =
+    Wrapped(values.toJSArray)
+
+  def apply[A](chain1: Chain[A], chain2: Chain[A]): Chain[A] =
+    Combine(chain1, chain2)
+
+  def empty[A]: Chain[A] =
+    Chain.Empty
+
+  def combineAll[A](chains: Chain[A]*): Chain[A] =
+    chains.toList.foldLeft(Chain.empty[A])(_ ++ _)
+
+  case object Empty extends Chain[Nothing]:
+    val isEmpty: Boolean                  = true
+    def map[B](f: Nothing => B): Chain[B] = this
+    def foreach(f: Nothing => Unit): Unit = ()
+    def toJSArray[B]: js.Array[B]         = js.Array[B]()
+
+    @SuppressWarnings(Array("scalafix:DisableSyntax.throw"))
+    def head: Nothing               = throw new NoSuchElementException("Chain.Empty.head")
+    def headOption: Option[Nothing] = None
+
+    override def equals(that: Any): Boolean =
+      that.isInstanceOf[Chain.Empty.type]
+
+  final case class Singleton[A](value: A) extends Chain[A]:
+    val isEmpty: Boolean               = false
+    def map[B](f: A => B): Chain[B]    = this.copy(value = f(value))
+    def foreach(f: A => Unit): Unit    = f(value)
+    def head: A                        = value
+    def headOption: Option[A]          = Some(value)
+    def toJSArray[B >: A]: js.Array[B] = js.Array[B](value)
+
+    override def equals(that: Any): Boolean =
+      given CanEqual[Empty.type, Any]   = CanEqual.derived
+      given CanEqual[Singleton[_], Any] = CanEqual.derived
+      given CanEqual[Combine[_], Any]   = CanEqual.derived
+      given CanEqual[Wrapped[_], Any]   = CanEqual.derived
+      given CanEqual[A, A]              = CanEqual.derived
+
+      try
+        that match
+          case Empty =>
+            false
+
+          case Singleton(v) =>
+            value == v.asInstanceOf[A]
+
+          case c @ Combine(_, _) =>
+            val cc = c.compact
+            cc.size == 1 && cc.head.asInstanceOf[A] == value
+
+          case Wrapped(arr) =>
+            arr.size == 1 && arr.head.asInstanceOf[A] == value
+
+      catch { _ => false }
+
+  final case class Combine[A](chain1: Chain[A], chain2: Chain[A]) extends Chain[A]:
+    val isEmpty: Boolean            = chain1.isEmpty && chain2.isEmpty
+    def map[B](f: A => B): Chain[B] = this.copy(chain1 = chain1.map(f), chain2 = chain2.map(f))
+    def foreach(f: A => Unit): Unit =
+      chain1.foreach(f)
+      chain2.foreach(f)
+    def head: A                        = chain1.head
+    def headOption: Option[A]          = chain1.headOption
+    def toJSArray[B >: A]: js.Array[B] = chain1.toJSArray ++ chain2.toJSArray
```

Comment:
There's a lot more room for optimization here, right? Essentially you should be able to allocate one array, and traverse the tree structure adding directly to that array.

Reply:
Funnily enough, that was the original implementation, then I changed it, and then I decided to stop fiddling until I could benchmark. Yes I think there's more to do.

Also the toJSArray benchmarks are currently totally unfair I realised afterwards as they just pass the original array back.

Anyway, it's been fun so far. 🙂

Reply:
> Also the toJSArray benchmarks are currently totally unfair I realised afterwards as they just pass the original array back.

Lol. They did look stupidly good 😆

Reply:
> Lol. They did look stupidly good 😆

Making stupid mistakes very publicly is my MO. Luckily I have no pride.

![image](https://user-images.githubusercontent.com/908709/166355655-99380d60-21f0-4353-91a2-9a5b843f7d3e.png)

That's more like it. Now we can see there's definitely work to do.

Reply:
Yeah so, interesting. The next problem is that if you want to do the trick of pushing into a single pre-allocated array, then you need to know the sizes, positions and offsets of everything. This moves us into Chunk/Rope territory of needing metadata of accumulated sizes or a binary tree of lengths. Exciting stuff!

Reply:
> Essentially you should be able to allocate one array, and traverse the tree structure adding directly to that array.

![image](https://user-images.githubusercontent.com/908709/167251868-c2942f24-c8a8-40c7-988c-c02681c329e2.png)

For the `map` operation (not `toJSArray`): In this table we have the original, then in the second two I'm allocating to a fixed array (one where I give it the size and one where I don't bother, the former is faster). Performance is worse however.

So it looks like for traversal operations, I can't do better than just looping the structure as efficiently as possible. It will always be slower than a native `map` or `foreach` - but I presume that's down to the megamorphic dispatch and additional allocation costs. I may well be missing something though?

That's not the end of the world. This structure at the moment is meant for fast construction, which it's doing just fine.

Reply:
Ah, but that depends on the depth and complexity of the tree. 🤔
On a complicated tree I'd be better off either doing the big array or flattening the whole thing first.